clash-lib/src/Clash/Core/HasFreeVars.hs

```diff
@@ -0,0 +1,50 @@
+{-|
+Copyright   : (C) 2021, QBayLogic B.V.
+License     : BSD2 (see the file LICENSE)
+Maintainer  : QBayLogic B.V. <devops@qbaylogic.com>
+
+Utility class to extract free variables from data which has variables.
+-}
+
+{-# LANGUAGE FlexibleInstances #-}
+
+module Clash.Core.HasFreeVars
+  ( HasFreeVars(..)
+  ) where
+
+import Data.Functor.Identity (Identity(..))
+
+import Clash.Core.FreeVars (localFVsOfTerms, tyFVsOfTypes)
+import Clash.Core.Term (Term)
+import Clash.Core.Type (Type)
+import Clash.Core.VarEnv
+
+class HasFreeVars a where
+  freeVarsOf :: a -> VarSet
+
+-- Identity is chosen as the Foldable thing here, as foldMap = coerce, which
+-- seems the best choice when you know there is only a single thing to fold.
+
+instance HasFreeVars Term where
+  {-# INLINE freeVarsOf #-}
+  freeVarsOf = localFVsOfTerms . Identity
+
+instance HasFreeVars Type where
+  {-# INLINE freeVarsOf #-}
+  freeVarsOf = tyFVsOfTypes . Identity
+
+instance (Foldable f, HasFreeVars a) => HasFreeVars (f a) where
+  {-# INLINE freeVarsOf #-}
+  freeVarsOf = foldMap freeVarsOf
+
+-- TODO I wonder if Clash.Core.FreeVars should have it's functions changed to
```

Comment:
I mean... if all uses of  `tyFVsOfTypes` and `localFVsOfTerms` are only within this module, then I see no reason to keep `tyFVsOfTypes` and `localFVsOfTerms`.

Reply:
Yeah, good point